Intro/README.md

```diff
@@ -0,0 +1,455 @@
+area cannot be changed after creating it.  One can delete and then
+recreate an object in order to "change" its public area, but this
+necessarily yields a new name (assuming no digest collisions).
+
+## Key Hierarchies
+
+TPMs have multiple key hierarchies, all rooted in a primary decrypt-only
+asymmetric private key derived from a seed, with arbitrarily complex
+trees of keys below the primary key:
+
+```
+                seed
+                 |
+                 |
+                 v
+     primary key (asymmetric encryption)
+                 |
+                 |
+                 v
+       secondary keys (of any kind)
+                 |
+                 |
+                 v
+                ...
+```
+
+There are three built-in hierarchies:
+
+ - platform hierarchy
+ - endorsement hierarchy
+ - storage hierarchy
+
+of which only the endorsement and storage hierarchies will be of
+interest to most readers.
+
+The endorsement hierarchy is used to authenticate (when needed) that a
+TPM is a legitimate TPM.  The primary endorsement key is known as the EK
+(endorsement key).  Hardware TPMs come with a certificate for the EK
+issued by the TPM's manufacturer.  This EK certificate ("EKcert") can be
+used to authenticate the TPM's legitimacy.  The EK's public key
+("EKpub") can be used to uniquely identify a TPM, and possibly link to
+the platform's, and even the platform's user(s)' identities.
+
+## Key Wrapping and Resource Management
+
+The primary key is always a decrypt-only asymmetric private key, and its
+corresponding public key is therefore encrypt-only.  This is largely
+because of key wrapping, where a symmetric key or asymmetric private key
+is encrypted to a TPM's EKpub so that it can be safely sent to that TPM
+so that that TPM can then decrypt and use that secret.
+
+As well as wrapping secrets by encryption to public keys, TPMs also use
+wrapping in a symmetric key known only to the TPM for the purpose of
+saving keys off the TPM.  This is used for resource management: since
+hardware TPMs have very limited resources, objects need to created or
+loaded, used, then saved off-TPM to make room for other objects to be
+loaded (unless they are not to be used again, then saving them is
+pointless).  Only a TPM that saved an object can load it again, but some
+objects can be exported to other TPMs by encrypting them to their
+destination TPMs' EKpubs.
+
+### Controlling Exportability of Keys
+
+A key that is `fixedTPM` cannot leave the TPM in cleartext.  It can be
+saved off the TPM it resides in, but only that TPM can load it again.
+
+A key that is `fixedParent` cannot be re-parented, though if its parent
+is neither `fixedParent` nor `fixedTPM` then the parent and its
+descendants can be moved as a group to some other TPM.
+
+## Persistence
+
+Cryptographic keys are, by default, not stored on non-volatile memory.
+Hardware TPMs have very little non-volatile (NV) memory.  They also have
```

Comment:
Does this mean that the key hierarchy described earlier gets reset on every boot by default? That seems surprising if so, and perhaps worthy of a bit more explanation.

Reply:
There are non-volatile keys and volatile keys, and the latter are "gone" on restart, but if you can save them then you can reload them post-restart.

Reply:
That's inaccurate.  It depends on the hierarchy.
The TCG uses the terms transient and persistent object.  I encourage you to use the TCG terms.  Using different terms leads to inaccuracies.

Reply:
I am also in favour of the terms transient and persistent because of two reasons

1. They are understandable as is
- transient - something temporary
- persistent - something that remains
2. Both terms seem to already have popularity among people who have tried or are using TPM

@kgold2 I would not mind having an explanation in the TPM intro about where exactly are persistent keys stored. AFAIK, it is in a segment of the non-volatile TPM memory. Perhaps, this is why @nicowilliams went with the terms volatile and non-volatile keys.

Reply:
> Perhaps, this is why @nicowilliams went with the terms volatile and non-volatile keys.

I'm still fairly new to this space, so I've not fully internalized the terminology, that's all.

> I don't understand 'a segment' in this context.

I don't think @tomoveu meant anything like "segment" in the sense of, say, segmented memory architectures.  I suspect what @tomoveu had in mind is that NV allocated for objects other than NV indexes is... does not get an NV index, so it's "separate" in some sense, though it can be a single pool of NV that is shared for all persistent objects _including_ NV indexes.

Reply:
> all persistent objects _including_ NV indexes.

NV Indexes are not persistent objects.  (NV Indexes are not objects).

I'm commenting about this because a tutorial that redefines TCG standard terms will lead to confusion.

The TPM specification has a glossary at the beginning of Part 1.